### PR TITLE
Add new `Rails/InverseOf` cop

### PR DESCRIPTION
**Pull request by bdewater**
This cop looks for has_(one|many) and belongs_to associations where ActiveRecord [can't automatically determine](https://github.com/rails/rails/blob/99f4d6e744285a1a634e6ba46cd500560573985c/activerecord/lib/active_record/reflection.rb#L569-L570) the inverse association because of a scope or the options used. This can result in unnecessary queries in some circumstances. `:inverse_of` must be manually specified for associations to work in both ways, or set to `false` to opt-out.

At Shopify we have tests that track SQL queries of critical paths, and fixing these associations resulted in less database calls.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by mikegee** [comment]
> At Shopify we have tests that track SQL queries of critical paths, and fixing these associations resulted in less database calls.

I'd like to understand the problem a bit more. Would you kindly show an example of an avoided database call?

**Comment by bdewater** [comment]
> I'd like to understand the problem a bit more.

These tests are there to ensure that developers do not accidentally add database queries, during our code review process any added SQL queries will have to be justified. This [2013 talk](https://youtu.be/j347oSSuNHA?t=16m55s) shows how the initial round of query optimization was done on paper by analyzing printed debug logs 😄 

> Would you kindly show an example of an avoided database call?

I initially started thinking about this PR after seeing the impact on Shopify's tests working on a feature that was adding a handful of queries in a case where we expected only one. After some investigation we fixed a misconfigured polymorphic association between classes that deal with orders and their various financial transactions. Now that the that the associations could be traversed in-memory _and_ with the new feature in place, we saw a net decrease of 3 and 5 model loads in two test cases. After that we found a few more instances, but these aren't covered by these query tests so I don't know the impact.

File: spec/rubocop/cop/rails/inverse_of_spec.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        RUBY
+      end
+    end
+
+    context 'Rails >= 5.2', :config do
+      let(:rails_version) { 5.2 }
+
+      it 'does not register an offense when not specifying `:inverse_of`' do
+        expect_no_offenses(
+          'belongs_to :foo, polymorphic: true'
```
Comment (bdewater) [inline spec/rubocop/cop/rails/inverse_of_spec.rb]: This is because of https://github.com/rails/rails/pull/28808

**Comment by Drenmi** [comment]
@mikegee: Check [this blog post](https://dev.mikamai.com/2013/11/15/you-should-always-specify-the-inverseof-option/). 🙂

**Comment by mikegee** [comment]
@Drenmi thanks! That has a clear example.

**Comment by bbatsov** [comment]
@bdewater Great first contribution! 👍 

P.S. You might also want to add an entry about this to the community Rails style guide https://github.com/bbatsov/rails-style-guide

**Comment by swrobel** [comment]
@bdewater can you clarify whether this cop is still accurate based on [your changes](https://github.com/rails/rails/pull/31446/files#diff-6cb426645499101965568d643d991110L738) to the rails docs? It would seem that `inverse_of` should only be required in the case that there's a scope, through, or foreign_key specified, but this cop seemingly barks any time I have any options specified on a has_many association.

**Comment by swrobel** [comment]
@bdewater upon further inspection, it seems to be barking because of my use of `class_name` which I [don't see listed in](https://github.com/rails/rails/blob/627ef06331b5c6c30e6100bfdaa38dbeb18c037a/activerecord/lib/active_record/reflection.rb#L573-L574) `INVALID_AUTOMATIC_INVERSE_OPTIONS`.

There is, however, [a spec for this cop](https://github.com/bbatsov/rubocop/commit/c3253dd0e54f4284eb591773db8c41fc54ed396e#diff-db636ffb09becd515822e3c106eea7b3R56) that was added by @wata727 that indicates that it should fail on `class_name`. Not sure why that was added as it's basically a footnote in that commit.

**Pull request merged**